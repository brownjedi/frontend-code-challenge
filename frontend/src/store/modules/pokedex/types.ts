export interface State {
  query: {
    search?: string
    filter?: {
      type?: string
      isFavorite?: boolean
    }
  }
}
