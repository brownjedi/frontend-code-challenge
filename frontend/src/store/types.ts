import { State as PokedexState } from './modules/pokedex'

export interface RootState {
  pokedex: PokedexState
}
