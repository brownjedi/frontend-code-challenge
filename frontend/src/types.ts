export interface PokemonFilterInput {
  type?: string
  isFavorite?: boolean
}

export interface PokemonsQueryInput {
  limit?: number
  offset?: number
  search?: string
  filter?: PokemonFilterInput
}

export interface PokemonsQueryResponse {
  pokemons: PokemonList
}

export interface PokemonByNameQueryResponse {
  pokemon: PokemonInfo
}

export interface PokemonTypesQueryResponse {
  pokemonTypes: PokemonTypes
}

export interface FavoriteQueryResponse {
  pokemon: Pick<PokemonOverview, 'id' | 'isFavorite'>
}

export interface UnFavoriteQueryResponse {
  pokemon: Pick<PokemonOverview, 'id' | 'isFavorite'>
}

export interface PokemonList {
  count: number
  edges: PokemonOverview[]
  limit: number
  offset: number
}

export interface PokemonOverview {
  id: string
  name: string
  image: string
  sprites: {
    normal: string
    animated: string
  }
  isFavorite: boolean
  classification: string
  types: string[]
}

export interface Evolution {
  id: string
  name: string
}

export interface RGBaster {
  backgroundColor: string
  textColor: string
}

export interface PokemonInfo extends PokemonOverview {
  number: number
  maxCP: number
  maxHP: number
  fleeRate: number
  resistant: string[]
  rgbaster: RGBaster
  sound: string
  weaknesses: string[]
  attacks: {
    special: {
      name: string
      type: string
      damage: number
    }
    fast: {
      name: string
      type: string
      damage: number
    }
  }[]
  evolutionRequirements: {
    name: string
    amount: number
  }
  evolutions: Evolution[]
  height: {
    minimum: string
    maximum: string
  }
  weight: {
    minimum: string
    maximum: string
  }
}

export interface PokemonTypes {
  pokemonTypes: string[]
}
