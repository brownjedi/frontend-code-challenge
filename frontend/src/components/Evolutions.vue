<template>
  <div :class="styles.container">
    <template :class="styles.evolution">
      <template v-for="(evolution, index) in evolutions" :key="evolution.id">
        <img :class="styles.arrow" src="/img/arrow--right.svg" alt="arrow" v-if="index !== 0" />
        <Evolution :id="evolution.id" :name="evolution.name" />
      </template>
    </template>
  </div>
</template>

<script lang="ts">
import { Evolution as EvolutionType } from '@/types'
import { defineComponent, PropType, useCssModule } from 'vue'
import Evolution from './Evolution.vue'

export default defineComponent({
  components: { Evolution },
  props: {
    evolutions: {
      type: Array as PropType<EvolutionType[]>,
    },
  },
  setup() {
    const styles = useCssModule()
    return { styles }
  },
})
</script>

<style lang="scss" module>
.container {
  display: flex;
  justify-content: center;
  align-items: center;
}

.evolution {
  display: flex;
  justify-content: center;
  align-items: baseline;

  .arrow {
    margin: 0 2rem;
    align-self: center;
  }

  @media (max-width: 64em) {
    flex-direction: column;
    align-items: center;
    .arrow {
      transform: rotate(90deg);
      margin: 2rem 0;
    }
  }
}
</style>
