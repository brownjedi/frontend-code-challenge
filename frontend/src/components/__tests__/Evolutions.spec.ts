import { shallowMount } from '@vue/test-utils'
import { Evolution as EvolutionType } from '@/types'
import Evolutions from '@/components/Evolutions.vue'

describe('components/Evolutions.vue', () => {
  it('has the expected html structure', () => {
    const evolutions: EvolutionType[] = [
      {
        id: '001',
        name: 'Bulbasaur',
      },
      {
        id: '002',
        name: 'Ivysaur',
      },
      {
        id: '003',
        name: 'Venusaur',
      },
    ]

    const wrapper = shallowMount(Evolutions, {
      props: {
        evolutions,
      },
    })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('should have n - 1 arrows in between for n evolutions', () => {
    const evolutions: EvolutionType[] = [
      {
        id: '001',
        name: 'Bulbasaur',
      },
      {
        id: '002',
        name: 'Ivysaur',
      },
      {
        id: '003',
        name: 'Venusaur',
      },
    ]

    const wrapper = shallowMount(Evolutions, {
      props: {
        evolutions,
      },
    })

    expect(wrapper.findAll('.arrow').length).toBe(evolutions.length - 1)
  })
})
