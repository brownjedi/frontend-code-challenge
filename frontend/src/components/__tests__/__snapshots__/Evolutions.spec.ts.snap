// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Evolutions.vue has the expected html structure 1`] = `
<div
  class="container"
>
  <template
    class="evolution"
  >
    
    
    <!--v-if-->
    <evolution-stub
      id="001"
      name="Bulbasaur"
    />
    
    
    <img
      alt="arrow"
      class="arrow"
      src="/img/arrow--right.svg"
    />
    <evolution-stub
      id="002"
      name="Ivysaur"
    />
    
    
    <img
      alt="arrow"
      class="arrow"
      src="/img/arrow--right.svg"
    />
    <evolution-stub
      id="003"
      name="Venusaur"
    />
    
    
  </template>
</div>
`;
